runtime/strings/bytes.cpp

```diff
@@ -0,0 +1,176 @@
+#include <gmp.h>
+#include <cstdlib>
+#include <cstring>
+#include <cstdint>
+#include <algorithm>
+#include <stdexcept>
+
+extern "C" {
+
+#define KCHAR char
+
+  struct blockheader {
+    uint64_t len;
+  };
+
+  struct string {
+    blockheader b;
+    KCHAR data[0];
+  };
+
+  mpz_ptr move_int(mpz_t);
+
+  string *hook_BYTES_empty() {
+    static string empty;
+    empty.b.len = 0;
+    return &empty;
+  }
+
+  string *hook_STRING_concat(string *a, string *b);
+  mpz_ptr hook_STRING_length(string *a);
+  string *hook_STRING_substr(string *a, mpz_t start, mpz_t end);
+
+  uint32_t getTagForSymbolName(const char *);
+
+  uint64_t tag_big_endian() {
+    return (((uint64_t)getTagForSymbolName("LblbigEndianBytes{}")) << 32) | 1;
+  }
+
+  uint64_t tag_unsigned() {
+    return (((uint64_t)getTagForSymbolName("LblunsignedBytes{}")) << 32) | 1;
+  }
+
+  mpz_ptr hook_BYTES_bytes2int(string *b, uint64_t endianness, uint64_t signedness) {
+    mpz_t result;
+    mpz_init(result);
+    int order = endianness == tag_big_endian() ? 1 : -1;
+    mpz_import(result, b->b.len, order, 1, 0, 0, b->data);
+    if (signedness != tag_unsigned() && b->b.len != 0) {
+      bool msb;
+      if (endianness == tag_big_endian()) {
+        msb = b->data[0] & 0x80;
+      } else {
+        msb = b->data[b->b.len - 1] & 0x80;
+      }
+      if (msb) {
+        mpz_t max;
+        mpz_init_set_ui(max, 1);
+        mpz_mul_2exp(max, max, b->b.len*8);
+        mpz_sub(result, result, max);
+        mpz_clear(max);
+      }
+    }
+    return move_int(result);
+  }
+
+  string *hook_BYTES_int2bytes(mpz_t len, mpz_t i, uint64_t endianness) {
+    if (!mpz_fits_ulong_p(len)) {
+      throw std::invalid_argument("Integer overflow on len");
+    }
+    unsigned long len_long = mpz_get_ui(len);
+    if (len_long == 0) {
+      return hook_BYTES_empty();
+    }
+    size_t sizeInBytes = (mpz_sizeinbase(i, 2) + 7) / 8;
+    bool neg = mpz_cmp_si(i, 0) < 0;
+    string *result = static_cast<string *>(malloc(sizeof(string) + len_long));
+    result->b.len = len_long;
+    memset(result->data, neg ? 0xff : 0x00, len_long);
+    int order = endianness == tag_big_endian() ? 1 : -1;
+    void *start = result->data + (endianness == tag_big_endian() ? len_long - sizeInBytes : 0);
+    mpz_t twos;
+    mpz_init_set(twos, i);
+    if (mpz_sgn(i) < 0) {
+      mpz_com(twos, twos);
+      mpz_add_ui(twos, twos, 1);
+    }
+    mpz_export(start, nullptr, order, 1, 0, 0, twos);
+    mpz_clear(twos);
+    return result;
+  }
+
+  string *hook_BYTES_bytes2string(string *b) {
+    size_t size = sizeof(string) + b->b.len;
+    string *result = static_cast<string *>(malloc(size));
+    memcpy(result, b, size);
+    return result;
+  }
```

Comment:
So this is essentially just copy.
I would throw an `inline` on this. It's likely the compiler will inline it, anyway, but I think it helps for readability.